### PR TITLE
[NFC] Explicitly pass a VFS when creating DiagnosticsEngine

### PR DESCRIPTION
**Pull request by kadircet**
Starting with 41e3919ded78d8870f7c95e9181c7f7e29aa3cc4 DiagnosticsEngine
creation might perform IO. It was implicitly defaulting to
getRealFileSystem. This patch makes it explicit by pushing the decision
making to callers.

It uses ambient VFS if one is available, and keeps using
`getRealFileSystem` if there aren't any VFS.

### CODE REVIEW

**Review by ilya-biryukov — APPROVED** [review]
LGTM, but please give others the time to chime in and approve.

File: clang/lib/Frontend/CompilerInstance.cpp [hunk 1]
```diff
@@ -332,23 +332,20 @@ static void SetupSerializedDiagnostics(DiagnosticOptions *DiagOpts,
...
+    DiagnosticConsumer *Client, bool ShouldOwnClient,
+    const CodeGenOptions *CodeGenOpts) {
   IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
-  IntrusiveRefCntPtr<DiagnosticsEngine>
-      Diags(new DiagnosticsEngine(DiagID, Opts));
-
-  if (!VFS)
-    VFS = llvm::vfs::getRealFileSystem();
+  IntrusiveRefCntPtr<DiagnosticsEngine> Diags(
+      new DiagnosticsEngine(DiagID, Opts));
```
Comment (bricknerb) [inline clang/lib/Frontend/CompilerInstance.cpp:348]: Why is this modified?
Is this just formatting?

File: clang/lib/Frontend/CompilerInstance.cpp [hunk 2]
```diff
@@ -368,11 +365,10 @@ IntrusiveRefCntPtr<DiagnosticsEngine> CompilerInstance::createDiagnostics(
     SetUpDiagnosticLog(Opts, CodeGenOpts, *Diags);
 
   if (!Opts->DiagnosticSerializationFile.empty())
-    SetupSerializedDiagnostics(Opts, *Diags,
-                               Opts->DiagnosticSerializationFile);
+    SetupSerializedDiagnostics(Opts, *Diags, Opts->DiagnosticSerializationFile);
```
Comment (bricknerb) [inline clang/lib/Frontend/CompilerInstance.cpp:368]: Just formatting?

**Review by bricknerb — APPROVED** [review]
Thanks for fixing!
It might be worth understanding whether this should have been caught by some test.
We might want to add a test or a check that would catch this.

**Comment by sylvestre** [comment]
I had to revert this for causing this issue with lldb:
https://github.com/llvm/llvm-project/issues/117145
